### PR TITLE
[minor] Commit pending status for PR deployment testing state

### PR DESCRIPTION
**Pull request by sunny299**
**What's the change or impact of this PR?**:

In case the PR deployment workflow trigger test pipeline on another branch or repository, we add the status `pending` in `Samsahai Deployment - Logs` job  while the Gitlab CI test pipeline is still running.
For increase visibility on Samsahai PR deployment workflow

 ---

- While testing `Samsahai Deployment - Logs` job display pending status

job-pending

- After test finished  `Samsahai Deployment - Logs` job display test result status 
job-finished

### CODE REVIEW

File: internal/util/gitlab/gitlab.go [hunk 1]
```diff
@@ -27,6 +27,8 @@ const (
 	CommitStatusSuccess CommitStatus = "success"
 	// CommitStatusFailure represents a failure of commit status
 	CommitStatusFailure CommitStatus = "failed"
+	// CommitStatusPending represents a failure of commit status
```
Comment (mmengspv) [inline internal/util/gitlab/gitlab.go:30]: Should be pending or not?
```suggestion
	// CommitStatusPending represents a pending of commit status
```

File: api/v1/queue_types.go [hunk 2]
```diff
@@ -341,6 +343,15 @@ func (qs *QueueStatus) IsConditionTrue(cond QueueConditionType) bool {
 	return false
 }
 
+func (qs *QueueStatus) IsConditionNull(cond QueueConditionType) bool {
```
Comment (mmengspv) [inline api/v1/queue_types.go:346]: Please change to IsNotContains or NotContains

File: internal/reporter/gitlab/reporter.go [hunk 3]
```diff
@@ -88,13 +88,30 @@ func (r *reporter) SendComponentUpgrade(configCtrl internal.ConfigController,
...
+			repository = b.Deployment.TestRunner.Gitlab.ProjectID
+		}
+	}
+
+	//prBundlName := ""
+	//if comp.PullRequestComponent != nil {
+	//	prBundlName = comp.PullRequestComponent.BundleName
+	//}
+
+	//repository := r.getGitlabRepository(teamName, prBundlName, configCtrl)
```
Comment (mmengspv) [inline internal/reporter/gitlab/reporter.go:113]: Please remove these comment

File: internal/reporter/gitlab/reporter.go [hunk 4]
```diff
@@ -88,13 +88,30 @@ func (r *reporter) SendComponentUpgrade(configCtrl internal.ConfigController,
...
+	conf, err := configCtrl.Get(teamName)
+	if err != nil {
+		return nil
+	}
+	repository := ""
+	for _, b := range conf.Spec.PullRequest.Bundles {
+		if b.Name == comp.Name {
+			repository = b.Deployment.TestRunner.Gitlab.ProjectID
+		}
+	}
```
Comment (mmengspv) [inline internal/reporter/gitlab/reporter.go:106]: Could we add a break, to end the loop if it founded?
```suggestion
	repository := ""
	for _, b := range conf.Spec.PullRequest.Bundles {
		if b.Name == comp.Name {
			repository = b.Deployment.TestRunner.Gitlab.ProjectID
			break
		}
	}
```

File: internal/reporter/gitlab/reporter.go [hunk 5]
```diff
@@ -171,9 +188,52 @@ func (r *reporter) SendPullRequestTriggerResult(configCtrl internal.ConfigContro
...
+	conf, err := configCtrl.Get(teamName)
+	if err != nil {
+		return nil
+	}
+	repository := ""
+	for _, b := range conf.Spec.PullRequest.Bundles {
+		if b.Name == prTestRunnerRpt.BundleName {
+			repository = b.Deployment.TestRunner.Gitlab.ProjectID
+		}
+	}
```
Comment (mmengspv) [inline internal/reporter/gitlab/reporter.go:210]: same as above

File: internal/reporter/gitlab/reporter.go [hunk 6]
```diff
@@ -88,13 +88,30 @@ func (r *reporter) SendComponentUpgrade(configCtrl internal.ConfigController,
...
 	if err != nil {
 		return nil
 	}
 
-	repository := r.getGitlabRepository(comp, configCtrl)
-	r.overrideGitlabCredential(comp, gitlabConfig)
+	conf, err := configCtrl.Get(teamName)
+	if err != nil {
+		return nil
+	}
```
Comment (Pohfy123) [inline internal/reporter/gitlab/reporter.go:100]: not sure is there any reason that we return nil if err != nil ?

Comment (Pohfy123) [reply internal/reporter/gitlab/reporter.go:100]: should we return err if team not found?

File: internal/reporter/gitlab/reporter.go [hunk 7]
```diff
@@ -171,9 +188,52 @@ func (r *reporter) SendPullRequestTriggerResult(configCtrl internal.ConfigContro
...
+
+	teamName := prTestRunnerRpt.TeamName
+	gitlabConfig, err := r.getGitlabConfig(teamName, configCtrl)
+	if err != nil {
+		return nil
+	}
+
+	conf, err := configCtrl.Get(teamName)
+	if err != nil {
+		return nil
```
Comment (Pohfy123) [inline internal/reporter/gitlab/reporter.go:203]: this one as well

File: internal/reporter/gitlab/reporter.go [hunk 8]
```diff
@@ -171,9 +188,52 @@ func (r *reporter) SendPullRequestTriggerResult(configCtrl internal.ConfigContro
 	return nil
 }
 
-func (r *reporter) post(gitlabConfig *s2hv1.ReporterGitlab,
-	repository, commitSHA, labelName, targetURL, description string,
-	commitStatus gitlab.CommitStatus, event internal.EventType) error {
+// SendPullRequestTestRunnerPendingResult implements the reporter SendPullRequestTestRunnerPendingResult function
+func (r *reporter) SendPullRequestTestRunnerPendingResult(configCtrl internal.ConfigController,
```
Comment (Pohfy123) [inline internal/reporter/gitlab/reporter.go:192]: This function is doing the thing similar to `SendPullRequestQueue`
Could we combine?

File: internal/reporter/gitlab/reporter.go [hunk 9]
```diff
@@ -213,22 +273,17 @@ func (r *reporter) getGitlabConfig(teamName string, configCtrl internal.ConfigCo
...
-	// no Gitlab configuration
-	if comp.PullRequestComponent == nil {
-		return ""
-	}
-
 	repository := ""
-	prBundleName := comp.PullRequestComponent.BundleName
-	if config.Status.Used.PullRequest != nil && len(config.Status.Used.PullRequest.Bundles) > 0 {
+	if config.Status.Used.PullRequest != nil && prBundleName != "" &&
+		len(config.Status.Used.PullRequest.Bundles) > 0 {
```
Comment (Pohfy123) [inline internal/reporter/gitlab/reporter.go:286]:
```suggestion
	if config.Status.Used.PullRequest != nil && 
	        prBundleName != "" &&
		len(config.Status.Used.PullRequest.Bundles) > 0 {
```

File: internal/samsahai/rpc.go [hunk 10]
```diff
@@ -287,6 +287,29 @@ func (c *controller) RunPostPullRequestTrigger(ctx context.Context, prTriggerRPC
...
+	*rpc.Empty, error) {
+
+	if err := c.authenticateRPC(ctx); err != nil {
+		return nil, err
+	}
+
+	prQueueName := teamWithPR.BundleName
+	prQueueNamespace := teamWithPR.Namespace
+	prQueue := &s2hv1.PullRequestQueue{}
+	err := c.client.Get(context.TODO(), types.NamespacedName{
```
Comment (Pohfy123) [inline internal/samsahai/rpc.go:300]: can we use the `ctx`?

File: internal/samsahai/rpc.go [hunk 11]
```diff
@@ -876,3 +899,30 @@ func (c *controller) sendPullRequestTriggerReport(prTrigger *s2hv1.PullRequestTr
...
+	configCtrl := c.GetConfigController()
+
+	teamComp := &s2hv1.Team{}
+	err := c.getTeam(teamWithPR.TeamName, teamComp)
+	if err != nil {
+		logger.Error(err, "cannot get team", "team", teamWithPR.TeamName)
+	}
+
+	if err := c.LoadTeamSecret(teamComp); err != nil {
+		logger.Error(err, "cannot load team secret", "team", teamComp.Name)
```
Comment (Pohfy123) [inline internal/samsahai/rpc.go:913]: curious that if we couldn't get team here, should we return the error? or we can continue?
won't it break the code after this e.g., `teamComp.Status.Used.Credential` ?

File: internal/staging/start_testrunner.go [hunk 12]
```diff
@@ -111,6 +124,8 @@ func (c *controller) startTesting(queue *s2hv1.Queue) error {
 	}
 	// test finished, change state to `s2hv1.Collecting`
 	if allTestFinished {
+
+
```
Comment (Pohfy123) [inline internal/staging/start_testrunner.go:128]: please remove extra newline

File: internal/staging/start_testrunner.go [hunk 13]
```diff
@@ -318,3 +333,45 @@ func (c *controller) setTestResultCondition(queue *s2hv1.Queue, testRunnerName s
 	// update queue back to k8s
 	return c.updateQueue(queue)
 }
+
+func (c *controller) sendTestPendingResult(queue *s2hv1.Queue) error {
+	headers := make(http.Header)
+	headers.Set(internal.SamsahaiAuthHeader, c.authToken)
+	ctx, err := twirp.WithHTTPRequestHeaders(context.TODO(), headers)
+	if err != nil {
+		logger.Error(err, "cannot set request header")
```
Comment (Pohfy123) [inline internal/staging/start_testrunner.go:342]: this one as well, shouldn't we return err?

File: internal/reporter/gitlab/reporter.go (hunk 8 above)
Comment (sunny299) [reply internal/reporter/gitlab/reporter.go:192]: it handles different parameter type, so i combined the duplicate step but still have separate functions

File: internal/reporter/gitlab/reporter.go [hunk 14]
```diff
@@ -171,9 +175,46 @@ func (r *reporter) SendPullRequestTriggerResult(configCtrl internal.ConfigContro
...
+
+	// send pull request log status pending while testrunner pipeline is running
+	err = r.post(gitlabConfig, projectID, commitSHA, LabelNameLogs, "", "",
+		gitlab.CommitStatusPending,
+		internal.PullRequestQueueType,
+	)
+	if err != nil {
+		return err
+	}
+	return nil
```
Comment (Pohfy123) [inline internal/reporter/gitlab/reporter.go:205]: can we just
```
return err
```

or 
return r.post(gitlabConfig, projectID, commitSHA, LabelNameLogs, "", "",
		gitlab.CommitStatusPending,
		internal.PullRequestQueueType,
	)

File: internal/staging/start_testrunner.go [hunk 15]
```diff
@@ -318,3 +331,46 @@ func (c *controller) setTestResultCondition(queue *s2hv1.Queue, testRunnerName s
...
+				"team", c.teamName, "component", queue.Spec.Name, "pr number", queue.Spec.PRNumber)
+			// set state, test pending status has been sent
+			queue.Status.SetCondition(
+				s2hv1.QueueTestPendingStatusSent,
+				v1.ConditionTrue,
+				"queue test pending status has been sent")
+			if err := c.updateQueue(queue); err != nil {
+				return err
+			}
+			return nil
```
Comment (Pohfy123) [inline internal/staging/start_testrunner.go:359]: can we add some error log?

File: internal/samsahai/rpc.go [hunk 16]
```diff
@@ -876,3 +901,33 @@ func (c *controller) sendPullRequestTriggerReport(prTrigger *s2hv1.PullRequestTr
...
+	prNumber := prQueue.Spec.PRNumber
+	commitSHA := prQueue.Spec.CommitSHA
+	for _, reporter := range c.reporters {
+		prTriggerRpt := s2h.NewPullRequestTestRunnerPendingReporter(c.configs, teamWithPR.TeamName,
+			bundleName, prNumber, commitSHA, teamComp.Status.Used.Credential)
+
+		if err := reporter.SendPullRequestTestRunnerPendingResult(configCtrl, prTriggerRpt); err != nil {
+			logger.Error(err, "cannot send pull request trigger result report",
+				"team", teamWithPR.TeamName, "bundle", bundleName, "prNumber", prNumber)
+		}
```
Comment (mmengspv) [inline internal/samsahai/rpc.go:930]: Could we return the err after log as well?

File: internal/staging/start_testrunner.go [hunk 17]
```diff
@@ -318,3 +331,46 @@ func (c *controller) setTestResultCondition(queue *s2hv1.Queue, testRunnerName s
...
+	// set state, cannot send test pending status
+	queue.Status.SetCondition(
+		s2hv1.QueueTestPendingStatusSent,
+		v1.ConditionFalse,
+		"cannot send queue test pending status")
+	if err := c.updateQueue(queue); err != nil {
+		return err
+	}
+
+	return err
```
Comment (Pohfy123) [inline internal/staging/start_testrunner.go:375]:
```
return c.updateQueue(queue)
```

Comment (Pohfy123) [reply internal/staging/start_testrunner.go:375]: or add log, same as the comment above, let's check first

**Comment by mmengspv** [comment]
LGTM

**Review by Pohfy123 — APPROVED** [review]

**Pull request merged**